src/runtime_src/core/include/experimental/xrt_xclbin.h

```diff
@@ -33,13 +36,270 @@
 typedef void* xrtXclbinHandle;
 
 #ifdef __cplusplus
+
+// Utility to replicate pimpl
+template<typename IMPL>
+class Pimpl
+{
+public:
+  Pimpl() = default;
+
+  virtual ~Pimpl() = default;
+
+  explicit Pimpl(std::shared_ptr<IMPL> handle) : handle(std::move(handle)) {}
+
+  std::shared_ptr<IMPL>
+  get_handle() const
+  {
+    return handle;
+  }
+
+protected:
+  std::shared_ptr<IMPL> handle;
+};
+
```

Comment:
This cannot be in the global namespace.   While I like the factoring, lets define this class in its own header file in the xrt::detail namespace.    For now do it in this header file, but in `namespace xrt { namespace detail {` and no P upper case.